### PR TITLE
Fixing SplunkPy fetch

### PR DESCRIPTION
**Pull request by guykeller**
## Status
- [ ] In Progress
- [x] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
fixes: https://github.com/demisto/etc/issues/35102
fixes: https://github.com/demisto/etc/issues/35805
fixes: https://github.com/demisto/etc/issues/34582
fixes: https://github.com/demisto/etc/issues/34529

## Description
Adding a look behind for the occurrence time + a deduping mechanism based on internally-defined and user-defined fields, to prevent missing events while not generating duplicate events.

## Screenshots
Paste here any images that will help the reviewer

## Minimum version of Cortex XSOAR
- [ ] 5.5.0
- [ ] 6.0.0
- [ ] 6.1.0
- [ ] 6.2.0

## Does it break backward compatibility?
   - [ ] Yes
       - Further details:
   - [ ] No

## Must have
- [ ] Tests
- [ ] Documentation

### CODE REVIEW

**Review by yuvalbenshalom — CHANGES_REQUESTED** [review]
see my comments

File: Packs/SplunkPy/Integrations/SplunkPy/SplunkPy.yml [hunk 1]
```diff
@@ -185,9 +185,27 @@ configuration:
...
+  name: unique_id_fields
+  type: 0
+  required: false
+  additionalinfo: Use this parameter to specify a list of coma separated fields, which
+    together are a unique identifier for the events you with to fetch
 description: Runs queries on Splunk servers.
-display: SplunkPy
-name: SplunkPy
+display: SplunkPy_post_conflicts
+name: SplunkPy_post_conflicts
```
Comment (guykeller) [inline Packs/SplunkPy/Integrations/SplunkPy/SplunkPy.yml:208]: will change prior to merge

**Review by yuvalbenshalom — APPROVED** [review]
Approved from my side. pending final review after fixes for the enrichment will be made.

**Comment by guykeller** [comment]
Adding `docs-approved` since the RN was taken from a fix by @amshamah419 in another PR

File: Packs/SplunkPy/Integrations/SplunkPy/SplunkPy.yml [hunk 2]
```diff
@@ -185,6 +185,24 @@ configuration:
...
+- display: Extensive logging (for debugging purposes)
+  name: extensive_logs
+  type: 8
+  required: false
+- display: Fetch backwards window for the events' occurrence time
+  name: occurrence_look_behind
+  defaultvalue: "15"
+  type: 0
+  required: false
+  additionalinfo: The fetch time range will be at least of the size specified here.
```
Comment (Itay4) [inline Packs/SplunkPy/Integrations/SplunkPy/SplunkPy.yml:197]: please add here to something like "change only if instructed by support"

File: Packs/SplunkPy/Integrations/SplunkPy/SplunkPy.yml [hunk 3]
```diff
@@ -185,6 +185,24 @@ configuration:
...
+    This will support events that have a gap between their occurrence time and their
+    index time in Splunk.
+- display: Unique ID fields
+  name: unique_id_fields
+  type: 0
+  required: false
+  additionalinfo: Use this parameter to specify a list of coma separated fields, which
+    together are a unique identifier for the events you with to fetch
```
Comment (Itay4) [inline Packs/SplunkPy/Integrations/SplunkPy/SplunkPy.yml:205]:
```suggestion
    together are a unique identifier for the events you with to fetch.
```

File: Packs/SplunkPy/Integrations/SplunkPy/SplunkPy.yml [hunk 4]
```diff
@@ -185,6 +185,24 @@ configuration:
   additionalinfo: The limit of how many events to retrieve per each one of the enrichment
     types (Drilldown, Asset, and Identity). To retrieve all events, enter "0" (not
     recommended).
+- display: Extensive logging (for debugging purposes)
```
Comment (Itay4) [inline Packs/SplunkPy/Integrations/SplunkPy/SplunkPy.yml:188]: add `Advanced: ` prefix to all displays

File: Packs/SplunkPy/Integrations/SplunkPy/SplunkPy.py [hunk 5]
```diff
@@ -88,15 +88,110 @@
...
+    extensive_log('ID after all fields were added: {}'.format(incident_custom_id))
+
+    unique_id = hashlib.md5(incident_custom_id).hexdigest()
+    extensive_log('Found incident ID is: {}'.format(unique_id))
+    return unique_id
+
+
+def extensive_log(message):
+    if demisto.params().get('extensive_logs', False):
+        demisto.info(message)
```
Comment (Itay4) [inline Packs/SplunkPy/Integrations/SplunkPy/SplunkPy.py:141]:
```suggestion
        demisto.debug(message)
```

File: Packs/SplunkPy/Integrations/SplunkPy/SplunkPy.py [hunk 6]
```diff
@@ -107,51 +202,137 @@ def fetch_notables(service, cache_object=None, enrich_notables=False):
...
+        notables.append(notable_incident)
+        inc = notable_incident.to_incident()
+        extensive_log('Incident data after parsing to notable: {}'.format(inc))
+        incident_id = create_incident_custom_id(inc)
+
+        if incident_id not in last_run_fetched_ids:
+            incident_ids_to_add.append(incident_id)
+            incidents.append(inc)
+        else:
+            extensive_log('SplunkPy - Dropped incident {} due to duplication.'.format(incident_id))
```
Comment (Itay4) [inline Packs/SplunkPy/Integrations/SplunkPy/SplunkPy.py:285]: add `[SplunkPy]` prefix to all logs

File: Packs/ExpanseV2/Playbooks/playbook-Expanse_Attribution.yml [hunk 7]
```diff
@@ -311,7 +311,7 @@ tasks:
                       iscontext: true
                     right:
                       value:
-                        simple: SplunkPy
+                        simple: SplunkPyPreRelease
```
Comment (Itay4) [inline Packs/ExpanseV2/Playbooks/playbook-Expanse_Attribution.yml:314]: why is this change?

File: Packs/SplunkPyPreRelease/Integrations/SplunkPyPreRelease/README.md [hunk 8]
```diff
@@ -0,0 +1,961 @@
...
+| fetch_time | The first timestamp to fetch in \<number\>\<time unit\> format. For example, "12 hours", "7 days", "3 months", "1 year". | False |
+| use_requests_handler | Use Python requests handler  | False |
+| type_field | Used only for mapping with the Select Schema option. The name of the field that contains the type of the event or alert. The default value is "source", which is a good option for notable events. However, you may choose any custom field that suits the need. | False |
+| use_cim | Use this option to get the mapping fields by Splunk CIM. See https://docs.splunk.com/Documentation/CIM/4.18.0/User/Overview for more info. | False | 
+| mirror_direction | Choose the direction to mirror the incident: Incoming (from Splunk to XSOAR), Outgoing (from XSOAR to Splunk), or Incoming and Outgoing (from/to SOAR and Splunk). | False |
+| close_incident | When selected, closing the Splunk notable event is mirrored in Cortex XSOAR. | False |
+| close_notable | When selected, closing the XSOAR incident is mirrored in Splunk. | False |
+| enabled_enrichments | The possible types of enrichment are: Drilldown, Asset, and Identity | False |
```
Comment (Itay4) [inline Packs/SplunkPyPreRelease/Integrations/SplunkPyPreRelease/README.md:48]: the new params are missing here

File: Packs/SplunkPyPreRelease/Integrations/SplunkPyPreRelease/SplunkPyPreRelease.yml [hunk 9]
```diff
@@ -0,0 +1,811 @@
...
+    This will support events that have a gap between their occurrence time and their
+    index time in Splunk.
+- display: Unique ID fields
+  name: unique_id_fields
+  type: 0
+  required: false
+  additionalinfo: Use this parameter to specify a list of comma separated fields, which
+    together are a unique identifier for the events you with to fetch.
+description: Runs queries on Splunk servers.
+display: SplunkPyPreRelease
```
Comment (Itay4) [inline Packs/SplunkPyPreRelease/Integrations/SplunkPyPreRelease/SplunkPyPreRelease.yml:207]:
```suggestion
display: SplunkPy Prerelease (Beta)
```

Comment (Itay4) [inline Packs/SplunkPyPreRelease/Integrations/SplunkPyPreRelease/SplunkPyPreRelease.yml:207]: add beta flag to this integration

File: Packs/SplunkPyPreRelease/Integrations/SplunkPyPreRelease/SplunkPyPreRelease.yml [hunk 10]
```diff
@@ -0,0 +1,811 @@
...
+  type: 0
+  required: false
+  additionalinfo: The limit of how many events to retrieve per each one of the enrichment
+    types (Drilldown, Asset, and Identity). To retrieve all events, enter "0" (not
+    recommended).
+- display: Extensive logging (for debugging purposes) - Please use this option unless advised otherwise.
+  name: extensive_logs
+  type: 8
+  required: false
+- display: 'Advanced (change only if advised bby support): Fetch backwards window for the events occurrence time'
```
Comment (Itay4) [inline Packs/SplunkPyPreRelease/Integrations/SplunkPyPreRelease/SplunkPyPreRelease.yml:192]:
```suggestion
- display: 'Advanced (change only if advised by support): Fetch backwards window for the events occurrence time'
```

File: Packs/SplunkPyPreRelease/Integrations/SplunkPyPreRelease/SplunkPyPreRelease.yml [hunk 11]
```diff
@@ -0,0 +1,811 @@
...
+    not enriched will be saved without the enrichment.
+- display: Number of Events Per Enrichment Type
+  name: num_enrichment_events
+  defaultvalue: "20"
+  type: 0
+  required: false
+  additionalinfo: The limit of how many events to retrieve per each one of the enrichment
+    types (Drilldown, Asset, and Identity). To retrieve all events, enter "0" (not
+    recommended).
+- display: Extensive logging (for debugging purposes) - Please use this option unless advised otherwise.
```
Comment (Itay4) [inline Packs/SplunkPyPreRelease/Integrations/SplunkPyPreRelease/SplunkPyPreRelease.yml:188]:
```suggestion
- display: Advanced: Extensive logging (for debugging purposes)
```

File: Packs/SplunkPyPreRelease/Integrations/SplunkPyPreRelease/SplunkPyPreRelease.yml [hunk 12]
```diff
@@ -0,0 +1,811 @@
...
+  required: false
+- display: 'Advanced (change only if advised bby support): Fetch backwards window for the events occurrence time'
+  name: occurrence_look_behind
+  defaultvalue: "15"
+  type: 0
+  required: false
+  additionalinfo: The fetch time range will be at least of the size specified here.
+    This will support events that have a gap between their occurrence time and their
+    index time in Splunk.
+- display: Unique ID fields
```
Comment (Itay4) [inline Packs/SplunkPyPreRelease/Integrations/SplunkPyPreRelease/SplunkPyPreRelease.yml:200]:
```suggestion
- display: Advanced: Unique ID fields
```

File: Packs/SplunkPyPreRelease/README.md [hunk 13]
```diff
@@ -0,0 +1,18 @@
+When a query or alert triggers a notable event (incident) in Splunk, the notable is stored in an index (table). An XSOAR analyst can fetch the notable event and its context data all from within the XSOAR console. 
+
+## What does this pack do?
+
+This pack is a pre -release version of the SplunkPy pack.
```
Comment (Itay4) [inline Packs/SplunkPyPreRelease/README.md:5]:
```suggestion
This pack is a pre-release version of the SplunkPy pack.
```

File: Packs/SplunkPyPreRelease/pack_metadata.json [hunk 14]
```diff
@@ -0,0 +1,48 @@
+{
+    "name": "SplunkPy_pre_release",
```
Comment (Itay4) [inline Packs/SplunkPyPreRelease/pack_metadata.json:2]:
```suggestion
    "name": "SplunkPy Prerelease",
```

File: Packs/SplunkPyPreRelease/pack_metadata.json [hunk 15]
```diff
@@ -0,0 +1,48 @@
...
+    "description": "Run queries on Splunk servers.",
```
Comment (Itay4) [inline Packs/SplunkPyPreRelease/pack_metadata.json:3]: please add short note to the desc about the prelease

File: Packs/SplunkPyPreRelease/Integrations/SplunkPyPreRelease/SplunkPyPreRelease_description.md [hunk 16]
```diff
@@ -0,0 +1,75 @@
+## SplunkPy
```
Comment (Itay4) [inline Packs/SplunkPyPreRelease/Integrations/SplunkPyPreRelease/SplunkPyPreRelease_description.md:1]: please add note here about the prelease

File: Packs/SplunkPyPreRelease/Integrations/SplunkPyPreRelease/SplunkPyPreRelease.yml [hunk 17]
```diff
@@ -0,0 +1,811 @@
...
+  ismappable: true
+  isremotesyncin: true
+  isremotesyncout: true
+  runonce: false
+  script: '-'
+  subtype: python2
+  type: python
+tests:
+- No tests (auto formatted)
+defaultclassifier: SplunkPyPreRelease
```
Comment (Itay4) [inline Packs/SplunkPyPreRelease/Integrations/SplunkPyPreRelease/SplunkPyPreRelease.yml:809]: are there such classifier and mapper

**Review by yuvalbenshalom — APPROVED** [review]

**Pull request merged**